### PR TITLE
Refactor how sections are stored in ELFOBJ ##refactor

### PR DESCRIPTION
**Pull request by ghost**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

* Fix some elf_notes problems (coverity)
* Add the flag is_valid to section information
* Refactor how section are stored
* Introduce a new misc function (rz_bin_elf_check_array)

The section loading use 2 steps because RzBinElfSection need shstrtab that need Elf_(Shdr).

PS: The next step is refactoring how strtab, dynstr, shstrab are stored

**Test plan**
`rz-test`

**Closing issues**

...

### CODE REVIEW

**Comment by XVilka** [comment]
Since we are removing `elf_write` here, what about removing it for all formats instead (in a separate PR)?

**Review by ret2libc — CHANGES_REQUESTED** [review]
Why are we removing elf_write?

If we remove that, `rz-bin -O` wont' work anymore, I think. There are some interesting things there, like changing entrypoints, change section permission.

Do we plan to drop all these?

**Comment by XVilka** [comment]
@ret2libc you can do this and much more with [LIEF](https://github.com/lief-project/LIEF). Our implementation is too basic, especially for resizing or moving sections. We could just create a plugin with LIEF in the future, IMHO.

**Comment by wargio** [comment]
i agree with Xvilka. LIEF is much much better than this. i would avoid to implement this since we would need to regenerate the structures and that is a mess.
We can add later a plugin via rz-pm to do this or an extra utils for this stuff.

**Comment by ret2libc** [comment]
I'm perfectly fine with removing the feature actually. But if we do, as @XVilka said, we should remove it for all plugins and we should also remove the flags/commands from rz-bin (`-O`) and rizin (`iO`).

**Comment by ghost** [comment]
~Fix 2 test in travis CI~ and create 2 new failure in github CI.

**Comment by ghost** [comment]
**2 failed tests**

I don't know why this 2 tests fail. On my computer it worked.

PS: I tested with a 32/64 bit binary

**Comment by ghost** [comment]
I believe, I fixed the issue (wasn't using the appropriate function to warn).

**Comment by ret2libc** [comment]
Are you going to remove the `write` method and the others bin_*_write plugins in a following PR?

File: librz/bin/format/elf/elf.c [hunk 1]
```diff
@@ -156,133 +156,63 @@ static void init_shdr_sdb(ELFOBJ *bin) {
...
 	if (!bin->shstrtab) {
-		perror("malloc");
 		return false;
 	}
 
-	int res = rz_buf_read_at(bin->b, bin->shstrtab_section->sh_offset, (ut8 *)bin->shstrtab, bin->shstrtab_section->sh_size);
+	int res = rz_buf_read_at(bin->b, offset, (ut8 *)bin->shstrtab, bin->shstrtab_size);
 	if (res < 0) {
-		bprintf("read (shstrtab) at 0x%" PFMT64x "\n", (ut64)bin->shstrtab_section->sh_offset);
+		bprintf("read (shstrtab) at 0x%" PFMT64x "\n", (ut64)offset);
```
Comment (wargio) [inline librz/bin/format/elf/elf.c]: what is bprintf? and since this is a fail, please use RZ_LOG_ERR here

File: librz/bin/format/elf/elf.c [hunk 2]
```diff
@@ -505,25 +421,29 @@ static bool rz_bin_elf_init(ELFOBJ *bin) {
...
 		}
 	} else {
-		if (!rz_bin_elf_init_shdr(bin)) {
-			bprintf("Cannot initialize section headers\n");
-		}
+		RzVector *sections = Elf_(rz_bin_elf_new_sections)(bin);
 
-		if (!rz_bin_elf_init_shstrtab(bin)) {
+		if (!rz_bin_elf_init_shstrtab(bin, sections)) {
 			bprintf("Cannot initialize strings table\n");
```
Comment (wargio) [inline librz/bin/format/elf/elf.c]: same here

File: librz/bin/format/elf/elf.c [hunk 3]
```diff
@@ -505,25 +421,29 @@ static bool rz_bin_elf_init(ELFOBJ *bin) {
...
 
-		if (!rz_bin_elf_init_dynstr(bin) && !Elf_(rz_bin_elf_is_relocatable)(bin)) {
-			bprintf("Cannot initialize dynamic strings\n");
-		}
-
-		bin->baddr = Elf_(rz_bin_elf_get_baddr)(bin);
-
 		if (!Elf_(rz_bin_elf_is_relocatable)(bin) && !Elf_(rz_bin_elf_is_static)(bin)) {
 			if (!rz_bin_elf_init_dynamic_section(bin) || !rz_bin_elf_init_strtab(bin)) {
 				bprintf("Cannot initialize dynamic section\n");
```
Comment (wargio) [inline librz/bin/format/elf/elf.c]: same here.

File: librz/bin/format/elf/elf.c [hunk 4]
```diff
@@ -505,25 +421,29 @@ static bool rz_bin_elf_init(ELFOBJ *bin) {
...
 			}
 		}
+
+		bin->baddr = Elf_(rz_bin_elf_get_baddr)(bin);
+
+		if (!rz_bin_elf_init_shdr(bin, sections)) {
+			bprintf("Cannot initialize section headers\n");
```
Comment (wargio) [inline librz/bin/format/elf/elf.c]: same here.

File: librz/bin/format/elf/elf.c [hunk 5]
```diff
@@ -505,25 +421,29 @@ static bool rz_bin_elf_init(ELFOBJ *bin) {
...
+		}
+
+		rz_vector_free(sections);
+
+		if (!rz_bin_elf_init_dynstr(bin) && !Elf_(rz_bin_elf_is_relocatable)(bin)) {
+			bprintf("Cannot initialize dynamic strings\n");
```
Comment (wargio) [inline librz/bin/format/elf/elf.c]: same here

**Review by wargio — CHANGES_REQUESTED** [review]

**Comment by ghost** [comment]
> Are you going to remove the `write` method and the others bin_*_write plugins in a following PR?

~~Yes, It is the idea.~~ I will send a PR that remove every RzBinWrite.

File: librz/bin/format/elf/elf.c (hunk 1 above)
Comment (ghost) [reply librz/bin/format/elf/elf.c]: It was a think I wanted to remove. It is a macro that printf if bin->verbose == true

**Comment by ghost** [comment]
@wargio Should I remove bin->verbose option?

**Comment by wargio** [comment]
> @wargio Should I remove bin->verbose option?

what is used for?

**Comment by ghost** [comment]
> > @wargio Should I remove bin->verbose option?
> 
> what is used for?

In some place to print verbose output. There isn't a lot of use.

**Comment by wargio** [comment]
> > > @wargio Should I remove bin->verbose option?
> > 
> > 
> > what is used for?
> 
> In some place to print verbose output. There isn't a lot of use.

Ok, remove the usage of this from the ELF code, and convert any eprintf/printf that uses that into the appropriate RZ_LOG_* counterpart.

Please check the output when loading some elf files, to see how verbose it is with the default log level (which is warning).

**Comment by ghost** [comment]
> > > > @wargio Should I remove bin->verbose option?
> > > 
> > > 
> > > what is used for?
> > 
> > 
> > In some place to print verbose output. There isn't a lot of use.
> 
> Ok, remove the usage of this from the ELF code, and convert any eprintf/printf that uses that into the appropriate RZ_LOG_* counterpart.
> 
> Please check the output when loading some elf files, to see how verbose it is with the default log level (which is warning).

I remove the use of bprintf when the information was not important for an end user.

**Review by XVilka — APPROVED** [review]

**Review by wargio — APPROVED** [review]

**Review by ret2libc — APPROVED** [review]

**Pull request merged**